Source/UltraRail/Private/WorldGen/WaveFunctionCollapse/BasicFieldObserver.cpp

```diff
@@ -0,0 +1,171 @@
+	FieldDepth = WaveCollapseGen->GetGenerationFieldDepth();
+	
+	const auto RegisteredIds = BiomeBlocks->RegisteredIds();
+	TArray<FBlockIdWeight> Weights = {};
+	for (const auto& RegisteredID : RegisteredIds)
+		Weights.Add({RegisteredID, 1.f / RegisteredIds.Num()});
+	
+	FieldState.Init({FCellState::Empty_State, Weights}, FieldWidth * FieldDepth);
+}
+
+bool ABasicFieldObserver::GetCurrentOptimalLocation_Implementation(int32& X, int32& Y)
+{
+	const auto& FieldSize = FieldState.Num();
+	int64 BestIndex = -1;
+	int64 LowestEntropy = BiomeBlockCount + 1;
+
+	for (int64 i = 0; i < FieldSize; i++)
+	{
+		const auto& [BlockID, Entropy] = FieldState[i];
+		if (BlockID != FCellState::Empty_State)
+			continue;
+
+		const auto& IndexedEntropy = Entropy.Num();
+		if (IndexedEntropy < LowestEntropy)
+		{
+			LowestEntropy = IndexedEntropy;
+			BestIndex = i;
+		}
+
+		// When the entropy is 1. We can immediately break out the loop as there can not be a lower entropy
+		if (LowestEntropy == 1)
+			break;
+	}
+
+	if (BestIndex == -1)
+	{
+		X = -1;
+		Y = -1;
+		return false;
+	}
+
+	TranslateIndexToCart_Implementation(BestIndex, X, Y);
+	return true;
+}
+
+void ABasicFieldObserver::GetFieldState_Implementation(TArray<FCellState>& Field)
+{
+	Field = FieldState;
+}
+
+bool ABasicFieldObserver::GetCell_Implementation(
+	const int32 X, const int32 Y, FCellState& CellState)
+{
+	if (X < 0 || X >= FieldWidth || Y < 0 || Y >= FieldDepth)
+		return false;
+	
+	CellState = FieldState[TranslateIndexFromCart_Implementation(X, Y)];
+	return true;
+}
+
+bool ABasicFieldObserver::SetCell_Implementation(const int32 X, const int32 Y, const FCellState& NewCellState)
+{
+	if (X < 0 || X >= FieldWidth || Y < 0 || Y >= FieldDepth)
+		return false;
+	
+	const auto Index = TranslateIndexFromCart_Implementation(X, Y);
+	FieldState[Index] = NewCellState;
+	return true;
+}
+
+int32 ABasicFieldObserver::GetFieldWidth_Implementation()
+{
+	return FieldWidth;
+}
+
+int32 ABasicFieldObserver::GetFieldDepth_Implementation()
+{
+	return FieldDepth;
+}
+
+bool ABasicFieldObserver::GetTopNeighbour_Implementation(const int32 X, const int32 Y, FCellState& TopNeighbour)
+{
+	return GetCell_Implementation(X, Y + 1, TopNeighbour);
+}
+
+bool ABasicFieldObserver::GetRightNeighbour_Implementation(int32 X, int32 Y, FCellState& RightNeighbour)
+{
+	return GetCell_Implementation(X + 1, Y, RightNeighbour);
+}
+
+bool ABasicFieldObserver::GetBottomNeighbour_Implementation(int32 X, int32 Y, FCellState& BottomNeighbour)
+{
+	return GetCell_Implementation(X, Y - 1, BottomNeighbour);
+}
+
+bool ABasicFieldObserver::GetLeftNeighbour_Implementation(int32 X, int32 Y, FCellState& LeftNeighbour)
+{
+	return GetCell_Implementation(X - 1, Y , LeftNeighbour);
+}
+
+bool ABasicFieldObserver::GetColumn_Implementation(const int32 ColumnIndex, TArray<FCellState>& Column)
+{
+	if (ColumnIndex >= FieldWidth || ColumnIndex < 0)
+		return false;
+	
+	Column.Init({FCellState::Empty_State, {}}, FieldDepth);
```

Comment:
Hier kan je beter gebruik maken van 'Array::Reserve()' aangezien die de memory daadwerkelijk reservt en 'Array::Init()' voegt elementen toe zonder te reserven

Reply:
Ik had niet helemaal verder gekeken dan dat mijn neus lang is, maar, 'Array::Init()' maakt wel onnodige copies